ufl/corealg/dag_traverser.py

```diff
@@ -0,0 +1,120 @@
+"""Base class for dag traversers."""
+
+from abc import ABC, abstractmethod
+from functools import singledispatchmethod, wraps
+
+from ufl.classes import Expr
+
+
+class DAGTraverser(ABC):
+    """Base class for dag traversers."""
+
+    def __init__(self, compress=True, vcache=None, rcache=None) -> None:
+        """Initialise."""
+        self._compress = compress
+        # Temporary data structures
+        # expr -> r = function(expr,...),  cache of intermediate results
+        self._vcache = {} if vcache is None else vcache
+        # r -> r,  cache of result objects for memory reuse
+        self._rcache = {} if rcache is None else rcache
+
+    def __call__(self, node: Expr, *args) -> Expr:
+        """Perform memoised DAG traversal with ``process`` singledispatch method.
+
+        Args:
+            node: `Expr` to start DAG traversal from.
+            args: arguments to the ``process`` singledispatchmethod.
+
+        Returns:
+            Processed `Expr`.
+
+        """
+        cache_key = (node, *args)
+        try:
+            return self._vcache[cache_key]
+        except KeyError:
+            result = self.process(node, *args)
+            # Optionally check if r is in rcache, a memory optimization
+            # to be able to keep representation of result compact
+            if self._compress:
+                result_ = self._rcache.get(result)
+                if result_ is None:
+                    # Cache miss: store in rcache
+                    self._rcache[result] = result
+                else:
+                    # Cache hit: Use previously computed object result_,
+                    # allowing result to be garbage collected as soon as possible
+                    result = result_
+            # Store result in cache
+            self._vcache[cache_key] = result
+            return result
+
+    @singledispatchmethod
+    @abstractmethod
+    def process(self, node: Expr, *args) -> Expr:
+        """Process node by type.
+
+        Args:
+            node: `Expr` to start DAG traversal from.
+            args: arguments to the ``process`` singledispatchmethod.
+
+        Returns:
+            Processed `Expr`.
+
+        """
+
+    def reuse_if_untouched(self, o: Expr) -> Expr:
+        """Reuse if touched.
+
+        Args:
+            o: `Expr` to start DAG traversal from.
+            new_ufl_operands: new ufl_operands of ``o``.
+
+        Returns:
+            Processed `Expr`.
+
+        """
+        new_ufl_operands = [self(operand) for operand in o.ufl_operands]
+        if all(nc == c for nc, c in zip(new_ufl_operands, o.ufl_operands)):
+            return o
+        else:
+            return o._ufl_expr_reconstruct_(*new_ufl_operands)
+
+    @staticmethod
+    def postorder(method):
+        """Suppress processed operands in arguments.
```

Comment:
Could this be clearer?

Reply:
I found it hard to keep it short. Please let me know if you have a suggestion. We at least have an expanded explanation right below.

Reply:
Changed to "Postorder decorator".